File: src/runtime/boxes/BbNotany.java
/**
 * BbNotany.java — NOTANY: match one char if NOT in set
 * Port of bb_notany.c / bb_notany.s
 *
 *   NOTANY_α:  if (Δ>=Ω || chars.indexOf(Σ[Δ])>=0)  goto NOTANY_ω;
 *              NOTANY=spec(Σ+Δ,1); Δ++;               goto NOTANY_γ;
 *   NOTANY_β:  Δ--;                                   goto NOTANY_ω;
 */
class BbNotany extends BbBox {
    private final String chars;

    public BbNotany(MatchState ms, String chars) { super(ms); this.chars=chars; }

    @Override public Spec alpha() {
        if (ms.delta >= ms.omega) return null;
        if (chars.indexOf(ms.sigma.charAt(ms.delta)) >= 0) return null;
        return new Spec(ms.delta++, 1);
    }

    @Override public Spec beta() { ms.delta--; return null; }
}

File: src/runtime/boxes/bb_notany.c
/* _XNNYC    NOTANY      match one char if NOT in set */
#include "bb_box.h"
#include <stdlib.h>
#include <string.h>


spec_t bb_notany(void *zeta, int entry)
{
    notany_t *ζ = zeta;
    spec_t NOTANY;
    if (entry==α)                                                               goto NOTANY_α;
    if (entry==β)                                                               goto NOTANY_β;
    NOTANY_α:       if (Δ>=Ω || strchr(ζ->chars,Σ[Δ]))                          goto NOTANY_ω;
                    NOTANY = spec(Σ+Δ,1); Δ++;                                  goto NOTANY_γ;
    NOTANY_β:       Δ--;                                                        goto NOTANY_ω;
    NOTANY_γ:                                                                   return NOTANY;
    NOTANY_ω:                                                                   return spec_empty;
}

notany_t *bb_notany_new(const char *chars)
{ notany_t *ζ=calloc(1,sizeof(notany_t)); ζ->chars=chars; return ζ; }
